File: src/p0001_two_sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        seen: dict[int, int] = {}
        for i, num in enumerate(nums):
            j = seen.get(target - num)
            if j is not None:
                return [j, i]

            seen[num] = i


def test_0001():
    s = Solution()

    # Example 1:
    assert s.twoSum([2, 7, 11, 15], 9) == [0, 1]

    # Example 2:
    assert s.twoSum([3, 2, 4], 6) == [1, 2]

    # Example 3:
    assert s.twoSum([3, 3], 6) == [0, 1]

File: src/p0001_two_sum.cpp
#include <catch2/catch_test_macros.hpp>
#include <stdexcept>
#include <unordered_map>
#include <vector>

namespace P0001 {
    class Solution {
      public:
        std::vector<int> twoSum(const std::vector<int> &nums, int target) {
            std::unordered_map<int, int> seen{{nums[0], 0}};
            for (auto it = nums.begin() + 1; it != nums.end(); ++it) {
                const auto pos = seen.find(target - *it);
                if (pos != seen.end()) {
                    return {pos->second, static_cast<int>(it - nums.begin())};
                }
                seen.insert({*it, static_cast<int>(it - nums.begin())});
            }
            throw std::runtime_error("No solution");
        }
    };

    TEST_CASE("0001 - Two Sum", "[Arrays][Hash Table]") {
        SECTION("Example 1") {
            Solution s;
            std::vector<int> nums = {2, 7, 11, 15};
            REQUIRE(s.twoSum(nums, 9) == std::vector<int>({0, 1}));
        }
        SECTION("Example 2") {
            Solution s;
            std::vector<int> nums = {3, 2, 4};
            REQUIRE(s.twoSum(nums, 6) == std::vector<int>({1, 2}));
        }
        SECTION("Example 3") {
            Solution s;
            std::vector<int> nums = {3, 3};
            REQUIRE(s.twoSum(nums, 6) == std::vector<int>({0, 1}));
        }
    }
} // namespace P0001
